File: AuthApi/Services/BlogService.cs
using AuthApi.Data;
using AuthApi.Models;
using Microsoft.EntityFrameworkCore;

namespace AuthApi.Services
{
    public class BlogService : IBlogService
    {
        private readonly ApplicationDbContext _context;

        public BlogService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<List<BlogPostDto>> GetAllPublishedPostsAsync()
        {
            var posts = await _context.BlogPosts
                .Where(p => p.Status == BlogPostStatus.Published)
                .OrderByDescending(p => p.PublishedAt ?? p.CreatedAt)
                .Select(p => new BlogPostDto
                {
                    Id = p.Id,
                    Title = p.Title,
                    Content = p.Content,
                    Summary = p.Summary,
                    AuthorId = p.AuthorId,
                    AuthorName = p.AuthorName,
                    Status = p.Status,
                    ViewCount = p.ViewCount,
                    CommentCount = p.CommentCount,
                    CreatedAt = p.CreatedAt,
                    UpdatedAt = p.UpdatedAt,
                    PublishedAt = p.PublishedAt
                })
                .ToListAsync();

            return posts;
        }

        public async Task<List<BlogPostDto>> GetAllPostsAsync()
        {
            var posts = await _context.BlogPosts
                .OrderByDescending(p => p.CreatedAt)
                .Select(p => new BlogPostDto
                {
                    Id = p.Id,
                    Title = p.Title,
                    Content = p.Content,
                    Summary = p.Summary,
                    AuthorId = p.AuthorId,
                    AuthorName = p.AuthorName,
                    Status = p.Status,
                    ViewCount = p.ViewCount,
                    CommentCount = p.CommentCount,
                    CreatedAt = p.CreatedAt,
                    UpdatedAt = p.UpdatedAt,
                    PublishedAt = p.PublishedAt
                })
                .ToListAsync();

            return posts;
        }

        public async Task<BlogPostDto?> GetPostByIdAsync(string id)
        {
            var post = await _context.BlogPosts
                .Where(p => p.Id == id)
                .Select(p => new BlogPostDto
                {
                    Id = p.Id,
                    Title = p.Title,
                    Content = p.Content,
                    Summary = p.Summary,
                    AuthorId = p.AuthorId,
                    AuthorName = p.AuthorName,
                    Status = p.Status,
                    ViewCount = p.ViewCount,
                    CommentCount = p.CommentCount,
                    CreatedAt = p.CreatedAt,
                    UpdatedAt = p.UpdatedAt,
                    PublishedAt = p.PublishedAt
                })
                .FirstOrDefaultAsync();

            return post;
        }

        public async Task<BlogPostDto> CreatePostAsync(BlogPostCreateDto createDto, string authorId, string authorName)
        {
            var post = new BlogPost
            {
                Id = Guid.NewGuid().ToString(),
                Title = createDto.Title,
                Content = createDto.Content,
                Summary = createDto.Summary,
                AuthorId = authorId,
                AuthorName = authorName,
                Status = createDto.Status,
                CreatedAt = DateTime.UtcNow
            };

            if (createDto.Status == BlogPostStatus.Published)
            {
                post.PublishedAt = DateTime.UtcNow;
            }

            _context.BlogPosts.Add(post);
            await _context.SaveChangesAsync();

            return new BlogPostDto
            {
                Id = post.Id,
                Title = post.Title,
                Content = post.Content,
                Summary = post.Summary,
                AuthorId = post.AuthorId,
                AuthorName = post.AuthorName,
                Status = post.Status,
                ViewCount = post.ViewCount,
                CommentCount = post.CommentCount,
                CreatedAt = post.CreatedAt,
                UpdatedAt = post.UpdatedAt,
                PublishedAt = post.PublishedAt
            };
        }

        public async Task<BlogPostDto?> UpdatePostAsync(string id, BlogPostUpdateDto updateDto, string userId)
        {
            var post = await _context.BlogPosts.FindAsync(id);
            if (post == null) return null;

            // Check if user is staff or the author
            var user = await _context.Users.FindAsync(userId);
            if (user == null || (user.Role != "staff" && post.AuthorId != userId))
            {
                return null;
            }

            var wasPublished = post.Status == BlogPostStatus.Published;
            
            post.Title = updateDto.Title;
            post.Content = updateDto.Content;
            post.Summary = updateDto.Summary;
            post.Status = updateDto.Status;
            post.UpdatedAt = DateTime.UtcNow;

            // Set published date if changing from draft to published
            if (!wasPublished && updateDto.Status == BlogPostStatus.Published)
            {
                post.PublishedAt = DateTime.UtcNow;
            }

            await _context.SaveChangesAsync();

            return new BlogPostDto
            {
                Id = post.Id,
                Title = post.Title,
                Content = post.Content,
                Summary = post.Summary,
                AuthorId = post.AuthorId,
                AuthorName = post.AuthorName,
                Status = post.Status,
                ViewCount = post.ViewCount,
                CommentCount = post.CommentCount,
                CreatedAt = post.CreatedAt,
                UpdatedAt = post.UpdatedAt,
                PublishedAt = post.PublishedAt
            };
        }

        public async Task<bool> DeletePostAsync(string id, string userId)
        {
            var post = await _context.BlogPosts.FindAsync(id);
            if (post == null) return false;

            // Check if user is staff or the author
            var user = await _context.Users.FindAsync(userId);
            if (user == null || (user.Role != "staff" && post.AuthorId != userId))
            {
                return false;
            }

            _context.BlogPosts.Remove(post);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> PublishPostAsync(string id, string userId)
        {
            var post = await _context.BlogPosts.FindAsync(id);
            if (post == null) return false;

            // Check if user is staff or the author
            var user = await _context.Users.FindAsync(userId);
            if (user == null || (user.Role != "staff" && post.AuthorId != userId))
            {
                return false;
            }

            post.Status = BlogPostStatus.Published;
            post.PublishedAt = DateTime.UtcNow;
            post.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> UnpublishPostAsync(string id, string userId)
        {
            var post = await _context.BlogPosts.FindAsync(id);
            if (post == null) return false;

            // Check if user is staff or the author
            var user = await _context.Users.FindAsync(userId);
            if (user == null || (user.Role != "staff" && post.AuthorId != userId))
            {
                return false;
            }

            post.Status = BlogPostStatus.Draft;
            post.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<BlogPostDto?> IncrementViewCountAsync(string id)
        {
            var post = await _context.BlogPosts.FindAsync(id);
            if (post == null) return null;

            post.ViewCount++;
            await _context.SaveChangesAsync();

            return new BlogPostDto
            {
                Id = post.Id,
                Title = post.Title,
                Content = post.Content,
                Summary = post.Summary,
                AuthorId = post.AuthorId,
                AuthorName = post.AuthorName,
                Status = post.Status,
                ViewCount = post.ViewCount,
                CommentCount = post.CommentCount,
                CreatedAt = post.CreatedAt,
                UpdatedAt = post.UpdatedAt,
                PublishedAt = post.PublishedAt
            };
        }

        public async Task<List<BlogPostDto>> SearchPostsAsync(string query)
        {
            var posts = await _context.BlogPosts
                .Where(p => p.Status == BlogPostStatus.Published && 
                           (p.Title.ToLower().Contains(query.ToLower()) || 
                            p.Content.ToLower().Contains(query.ToLower()) ||
                            p.Summary.ToLower().Contains(query.ToLower())))
                .OrderByDescending(p => p.PublishedAt ?? p.CreatedAt)
                .Select(p => new BlogPostDto
                {
                    Id = p.Id,
                    Title = p.Title,
                    Content = p.Content,
                    Summary = p.Summary,
                    AuthorId = p.AuthorId,
                    AuthorName = p.AuthorName,
                    Status = p.Status,
                    ViewCount = p.ViewCount,
                    CommentCount = p.CommentCount,
                    CreatedAt = p.CreatedAt,
                    UpdatedAt = p.UpdatedAt,
                    PublishedAt = p.PublishedAt
                })
                .ToListAsync();

            return posts;
        }

        public async Task<List<BlogCommentDto>> GetCommentsByPostIdAsync(string postId)
        {
            var comments = await _context.BlogComments
                .Where(c => c.BlogPostId == postId)
                .OrderBy(c => c.CreatedAt)
                .Select(c => new BlogCommentDto
                {
                    Id = c.Id,
                    BlogPostId = c.BlogPostId,
                    UserId = c.UserId,
                    UserName = c.UserName,
                    Content = c.Content,
                    CreatedAt = c.CreatedAt,
                    UpdatedAt = c.UpdatedAt
                })
                .ToListAsync();

            return comments;
        }

        public async Task<BlogCommentDto> CreateCommentAsync(string postId, BlogCommentCreateDto createDto, string userId, string userName)
        {
            var comment = new BlogComment
            {
                Id = Guid.NewGuid().ToString(),
                BlogPostId = postId,
                UserId = userId,
                UserName = userName,
                Content = createDto.Content,
                CreatedAt = DateTime.UtcNow
            };

            _context.BlogComments.Add(comment);

            // Update comment count
            var post = await _context.BlogPosts.FindAsync(postId);
            if (post != null)
            {
                post.CommentCount++;
            }

            await _context.SaveChangesAsync();

            return new BlogCommentDto
            {
                Id = comment.Id,
                BlogPostId = comment.BlogPostId,
                UserId = comment.UserId,
                UserName = comment.UserName,
                Content = comment.Content,
                CreatedAt = comment.CreatedAt,
                UpdatedAt = comment.UpdatedAt
            };
        }

        public async Task<bool> DeleteCommentAsync(string commentId, string userId)
        {
            var comment = await _context.BlogComments.FindAsync(commentId);
            if (comment == null) return false;

            // Check if user is staff or the comment author
            var user = await _context.Users.FindAsync(userId);
            if (user == null || (user.Role != "staff" && comment.UserId != userId))
            {
                return false;
            }

            _context.BlogComments.Remove(comment);

            // Update comment count
            var post = await _context.BlogPosts.FindAsync(comment.BlogPostId);
            if (post != null && post.CommentCount > 0)
            {
                post.CommentCount--;
            }

            await _context.SaveChangesAsync();
            return true;
        }
    }
}

File: src/services/blogService.ts
import authApi from './authApi';

// Types for blog
export interface BlogPost {
    id: string;
    title: string;
    content: string;
    summary: string;
    authorId: string;
    authorName: string;
    status: 0 | 1; // 0 = Draft, 1 = Published
    viewCount: number;
    commentCount: number;
    createdAt: string;
    updatedAt?: string;
    publishedAt?: string;
}

export interface BlogComment {
    id: string;
    blogPostId: string;
    userId: string;
    userName: string;
    content: string;
    createdAt: string;
    updatedAt?: string;
}

export interface BlogPostCreateDto {
    title: string;
    content: string;
    summary: string;
    status: 0 | 1; // 0 = Draft, 1 = Published
}

export interface BlogPostUpdateDto {
    title: string;
    content: string;
    summary: string;
    status: 0 | 1; // 0 = Draft, 1 = Published
}

export interface BlogCommentCreateDto {
    content: string;
}

// Blog Post API functions
export const getAllPublishedPosts = async (): Promise<BlogPost[]> => {
    try {
        const response = await authApi.get('/blog/published');
        return response.data;
    } catch (error) {
        console.error('Error fetching published posts:', error);
        return [];
    }
};

export const getAllPosts = async (): Promise<BlogPost[]> => {
    try {
        const response = await authApi.get('/blog/all');
        return response.data;
    } catch (error) {
        console.error('Error fetching all posts:', error);
        return [];
    }
};

export const getPostById = async (id: string): Promise<BlogPost | null> => {
    try {
        const response = await authApi.get(`/blog/${id}`);
        return response.data;
    } catch (error) {
        console.error('Error fetching post:', error);
        return null;
    }
};

export const incrementViewCount = async (id: string): Promise<BlogPost | null> => {
    try {
        const response = await authApi.post(`/blog/${id}/view`);
        return response.data;
    } catch (error) {
        console.error('Error incrementing view count:', error);
        return null;
    }
};

export const searchPosts = async (query: string): Promise<BlogPost[]> => {
    try {
        const response = await authApi.get(`/blog/search?query=${encodeURIComponent(query)}`);
        return response.data;
    } catch (error) {
        console.error('Error searching posts:', error);
        return [];
    }
};

export const createPost = async (postData: BlogPostCreateDto): Promise<BlogPost | null> => {
    try {
        const response = await authApi.post('/blog', postData);
        return response.data;
    } catch (error) {
        console.error('Error creating post:', error);
        return null;
    }
};

export const updatePost = async (id: string, postData: BlogPostUpdateDto): Promise<BlogPost | null> => {
    try {
        const response = await authApi.put(`/blog/${id}`, postData);
        return response.data;
    } catch (error) {
        console.error('Error updating post:', error);
        return null;
    }
};

export const deletePost = async (id: string): Promise<boolean> => {
    try {
        await authApi.delete(`/blog/${id}`);
        return true;
    } catch (error) {
        console.error('Error deleting post:', error);
        return false;
    }
};

export const publishPost = async (id: string): Promise<boolean> => {
    try {
        await authApi.post(`/blog/${id}/publish`);
        return true;
    } catch (error) {
        console.error('Error publishing post:', error);
        return false;
    }
};

export const unpublishPost = async (id: string): Promise<boolean> => {
    try {
        await authApi.post(`/blog/${id}/unpublish`);
        return true;
    } catch (error) {
        console.error('Error unpublishing post:', error);
        return false;
    }
};

// Blog Comment API functions
export const getCommentsByPostId = async (postId: string): Promise<BlogComment[]> => {
    try {
        const response = await authApi.get(`/blog/${postId}/comments`);
        return response.data;
    } catch (error) {
        console.error('Error fetching comments:', error);
        return [];
    }
};

export const createComment = async (postId: string, commentData: BlogCommentCreateDto): Promise<BlogComment | null> => {
    try {
        const response = await authApi.post(`/blog/${postId}/comments`, commentData);
        return response.data;
    } catch (error) {
        console.error('Error creating comment:', error);
        return null;
    }
};

export const deleteComment = async (commentId: string): Promise<boolean> => {
    try {
        await authApi.delete(`/blog/comments/${commentId}`);
        return true;
    } catch (error) {
        console.error('Error deleting comment:', error);
        return false;
    }
};

// Legacy functions for backward compatibility (will be removed)
export const addComment = async (postId: string, userName: string, content: string): Promise<BlogComment | null> => {
    return await createComment(postId, { content });
};
